modules/ml/src/main/java/org/apache/ignite/ml/optimization/GradientDescent.java

```diff
@@ -0,0 +1,194 @@
+ */
+
+package org.apache.ignite.ml.optimization;
+
+import org.apache.ignite.ml.math.Matrix;
+import org.apache.ignite.ml.math.Vector;
+import org.apache.ignite.ml.math.impls.matrix.SparseDistributedMatrix;
+import org.apache.ignite.ml.math.impls.vector.DenseLocalOnHeapVector;
+import org.apache.ignite.ml.math.impls.vector.FunctionVector;
+import org.apache.ignite.ml.optimization.util.SparseDistributedMatrixMapReducer;
+
+/**
+ * Gradient descent optimizer.
+ */
+public class GradientDescent {
+    /**
+     * Function which computes gradient of the loss function at any given point.
+     */
+    private final GradientFunction lossGradient;
+
+    /**
+     * Weights updater applied on every gradient descent step to decide how weights should be changed.
+     */
+    private final Updater updater;
+
+    /**
+     * Max number of gradient descent iterations.
+     */
+    private int maxIterations = 1000;
+
+    /**
+     * Convergence tolerance is condition which decides iteration termination.
+     */
+    private double convergenceTol = 1e-8;
+
+    /**
+     * New gradient descent instance based of loss function and updater.
+     *
+     * @param lossGradient Function which computes gradient of the loss function at any given point
+     * @param updater Weights updater applied on every gradient descent step to decide how weights should be changed
+     */
+    public GradientDescent(GradientFunction lossGradient, Updater updater) {
+        this.lossGradient = lossGradient;
+        this.updater = updater;
+    }
+
+    /**
+     * Sets max number of gradient descent iterations.
+     *
+     * @param maxIterations Max number of gradient descent iterations
+     * @return This gradient descent instance
+     */
+    public GradientDescent withMaxIterations(int maxIterations) {
+        if (maxIterations < 0)
+            throw new IllegalArgumentException("Number of iterations must be non-negative but got " + maxIterations);
+        this.maxIterations = maxIterations;
+        return this;
+    }
+
+    /**
+     * Sets convergence tolerance.
+     *
+     * @param convergenceTol Condition which decides iteration termination
+     * @return This gradient descent instance
+     */
+    public GradientDescent withConvergenceTol(double convergenceTol) {
+        if (convergenceTol < 0)
+            throw new IllegalArgumentException("Convergence tolerance must be non-negative but got " + convergenceTol);
+        this.convergenceTol = convergenceTol;
+        return this;
+    }
+
+    /**
+     * Computes point where loss function takes minimal value.
+     *
+     * @param data Inputs parameters of loss function
+     * @param initialWeights Initial weights
+     * @return Point where loss function takes minimal value
+     */
+    public Vector optimize(Matrix data, Vector initialWeights) {
+        Vector weights = initialWeights, oldWeights = null, oldGradient = null;
+        if (data instanceof SparseDistributedMatrix) {
```

Comment:
Could we refactor it to one method to avoid copy-paste?

Reply:
Good idea, done.